Document dev and test workflow using workspaces (#656)
navapbc/template-infra: docs/infra/develop-and-test-infrastructure-in-isolation-using-workspaces.md

This document describes a workflow that leverages [Terraform workspaces](https://developer.hashicorp.com/terraform/language/state/workspaces) for developing and testing infrastructure changes in isolation so that they can be tested and peer reviewed before being merged into main.

## Overview

By default, each Terraform root module has a single workspace named `default`. Workspaces allow you to deploy multiple instances of the root module configuration without configuring a new backend. When you run `terraform apply` in a separate workspace, a parallel set of infrastructure resources are created.

There are a few notable differences with resources created in a non-default workspace:

1. When using terraform, you cannot deploy the same resource with the same name. The `terraform apply` will fail with an error like "A resource with the ID already exists". Therefore, to avoid naming conflicts, we prefix the resource names with the workspace name.
2. Some resources, such as database and storage buckets, enable deletion protection to prevent accidental deletion. However, non-default workspaces are intended to be temporary, so deletion protection is disabled in non-default workspaces.

> 3. Resources that are difficult to create in isolation, such as [DNS records](https://github.com/navapbc/template-infra/blob/2cda6da18c84aa5a3dfb038ab32be4fac363af21/infra/modules/service/dns.tf#L3), are not created at all.

Reviewer: Are there any other resources like this? I think it would be useful to either list them here or have a keyword search term that we can use to find them all in the codebase.

Author: One possible keyword search term would be `is_temporary`. Are you suggesting that we tell readers to search for uses of `is_temporary` to find other examples?

Reviewer: Either that or to list them here in the documentation.

## Development workflow

Follow these steps if you want to develop and test a change to the service layer. Make the appropriate changes to the `-chdir` flag if you want to make a change to a different layer, such as the database layer or network layer.

### 1. Create a new workspace

First, make sure that the Terraform root module is initialized to the dev environment

```bash

> terraform -chdir=infra/$app_name/service init -backend-config=dev.s3.tfbackend

Reviewer: At some point, we should land on a documentation styleguide for consistency and approachability. I think all of the docs use `<APP_NAME>` to indicate fields that should be replaced by the person running the command.

Reviewer: Our Makefile executes the following `init` for `infra-update-app-service`:

```sh
terraform -chdir="infra/$(APP_NAME)/service" init -input=false -reconfigure -backend-config="$(ENVIRONMENT).s3.tfbackend"
```

Would it be more consistent to include the `-input=false` and `-reconfigure` flags in this instruction? Are there potential issues the user might run into without these flags?

Author: Good point on style guide I'll update.

`-reconfigure` is worth including but `-input=false` is only for running in a CI context not interactively.

```

> Then creat a new workspace. Since the workspace name is used to prefix resource names, use a short workspace name to avoid hitting resource name size limits. Assuming you're only working on one thing at a time (following the Kanban principle of limiting work in progress), your initials would make a good workspace name.

Reviewer 2: ```suggestion
Then create a new workspace. Since the workspace name is used to prefix resource names, use a short workspace name to avoid hitting resource name size limits. Assuming you're only working on one thing at a time (following the Kanban principle of limiting work in progress), your initials would make a good workspace name. For example, if your initials were LY, you would use this command:
```

Reviewer 2: (making it very clear that `ly` is a placeholder)

Author: Good point, I think I'll use a placeholder like <WORKSPACe_NAME> to be consistent with other docs.

> Then creat a new workspace. Since the workspace name is used to prefix resource names, use a short workspace name to avoid hitting resource name size limits. Assuming you're only working on one thing at a time (following the Kanban principle of limiting work in progress), your initials would make a good workspace name. For example, if your name was Loren Yu, you could use `ly` as your workspace name.

Reviewer: Typo: `Then create`

Reviewer: Nit: "resource name character limits" instead of "size limits".

```bash
terraform -chdir=infra/$app_name/service workspace new ly
```

> Verify that the new workspace was created and selected

Reviewer 2: ```suggestion
Verify that the new workspace was created and selected:
```

```bash
# List all workspaces, with a * next to the selected workspace
terraform -chdir=infra/$app_name/service workspace list
# - OR -
# Show your current selected workspace
terraform -chdir=infra/$app_name/service workspace show
```

### 2. Create resources in your workspace

```bash
terraform -chdir=infra/$app_name/service apply -var=environment_name=dev
# - OR -
make infra-update-app-service "APP_NAME=$app_name" ENVIRONMENT=dev
```

### 3. Clean up after merging to main and deploying changes to default workspace

Finally, after you merged your pull request and have deployed your changes to the default workspace, clean up your workspace so that you don't continue to accrue costs from the infrastructure resources.

```bash
# Destroy all infrastructure resources within the workspace
terraform -chdir=infra/$app_name/service destroy -var=environment_name=dev
# Select default workspace so that you can delete your workspace, since you can't delete the selected workspace
terraform -chdir=infra/$app_name/service workspace select default

> # Delete your workspace

Reviewer 2: ```suggestion
# Delete your workspace (replace `ly` with your workspace's name)
```

Author: I replaced with a placeholder
